### PR TITLE
CB-3013 alter Snowflake session parameter on start 

### PR DESCRIPTION
**Pull request by LonwoLonwo**
to avoid SELECT statement reading issues

2022-12-27 10_50_30-Window

Knowable issue for now: `SELECT CURRENT_DATABASE(), CURRENT_SCHEMA()` is working before `ALTER SESSION`, and because of that, we do not see the default catalog at the start (if it is not in the bootstrap settings).

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeDataSource.java [hunk 1]
```diff
@@ -76,6 +83,22 @@ protected JDBCExecutionContext createExecutionContext(JDBCRemoteInstance instanc
...
+    @Override
+    public void initialize(@NotNull DBRProgressMonitor monitor) throws DBException {
+        try (JDBCSession session = DBUtils.openMetaSession(monitor, this, "Alter Snowflake Session")) {
+            // JDBC Driver throws java.lang.NoClassDefFoundError for class RootAllocator in Arrow Library, while trying to run a select query.
+            // So this is official workaround for systems with Java 17 and if Arrow library installation is not possible
+            // or if the Operating System is not supported by Arrow.
+            try (JDBCStatement dbStat = session.createStatement()) {
+                dbStat.execute("ALTER SESSION SET JDBC_QUERY_RESULT_FORMAT='JSON'");
+            } catch (SQLException e) {
+                // Only logging because
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeDataSource.java]: ...because what?

**Review by ShadelessFox — APPROVED** [review]

File: plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeDataSource.java (hunk 1 above)
Comment (LonwoLonwo) [reply plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeDataSource.java]: ouch

**Comment by LonwoLonwo** [comment]
Any other ideas are welcome. Because I do not know - what if old Sbowlake databases do not support 
`ALTER SESSION SET JDBC_QUERY_RESULT_FORMAT='JSON'`

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/impl/AbstractExecutionContext.java [hunk 2]
```diff
@@ -119,13 +121,22 @@ protected boolean initContextBootstrap(@NotNull DBRProgressMonitor monitor, bool
...
         }
         return false;
     }
 
+    /**
+     * Returns a list of queries that should be performed before initializing the session.
+     */
+    @NotNull
+    public List<String> getExtraInitQueries() {
+        return new ArrayList<>();
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/impl/AbstractExecutionContext.java]:
```suggestion
        return List.of();
```

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeExecutionContext.java [hunk 3]
```diff
@@ -204,4 +206,13 @@ private void setActiveSchema(DBRProgressMonitor monitor, @NotNull String schemaN
...
         }
     }
+
+    @NotNull
+    @Override
+    public List<String> getExtraInitQueries() {
+        // JDBC Driver throws java.lang.NoClassDefFoundError for class RootAllocator in Arrow Library, while trying to run a select query.
+        // So this is official workaround for systems with Java 17 and if Arrow library installation is not possible
+        // or if the Operating System is not supported by Arrow.
+        return Collections.singletonList("ALTER SESSION SET JDBC_QUERY_RESULT_FORMAT='JSON'");
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeExecutionContext.java]: Use modern API

```suggestion
        return List.of("ALTER SESSION SET JDBC_QUERY_RESULT_FORMAT='JSON'");
```

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/impl/AbstractExecutionContext.java (hunk 2 above)
Comment (LonwoLonwo) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/impl/AbstractExecutionContext.java]: * Returns an unmodifiable list containing zero elements.

I do not think so.

Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/impl/AbstractExecutionContext.java]: Maybe you'd prefer `Collections.emptyList()`?

Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/impl/AbstractExecutionContext.java]: Both `List.of()` and `Collections.emptyList()` returns a static instance of an empty list, resulting in no additional allocations.

image

File: plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeExecutionContext.java [hunk 4]
```diff
@@ -204,4 +205,13 @@ private void setActiveSchema(DBRProgressMonitor monitor, @NotNull String schemaN
...
         }
     }
+
+    @NotNull
+    @Override
+    public List<String> getExtraInitQueries() {
+        // JDBC Driver throws java.lang.NoClassDefFoundError for class RootAllocator in Arrow Library, while trying to run a select query.
+        // So this is official workaround for systems with Java 17 and if Arrow library installation is not possible
+        // or if the Operating System is not supported by Arrow.
+        return List.of("ALTER SESSION SET JDBC_QUERY_RESULT_FORMAT='JSON'");
```
Comment (Destrolaric) [inline plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeExecutionContext.java]: Can we check if we are running under Java 17, and only then use this?

**Review by ShadelessFox — APPROVED** [review]

**Review by E1izabeth — APPROVED** [review]

**Comment by LonwoLonwo** [comment]
New code version again :D

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/impl/AbstractExecutionContext.java [hunk 5]
```diff
@@ -124,26 +113,31 @@ protected boolean initContextBootstrap(@NotNull DBRProgressMonitor monitor, bool
...
                 }
+                executeExtraInitQueries(session);
             }
             return true;
+        } else {
+            boolean executed;
+            try (DBCSession session = openSession(monitor, DBCExecutionPurpose.UTIL, "Run extra init queries")) {
+                executed = executeExtraInitQueries(session);
+            }
+            return executed;
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/impl/AbstractExecutionContext.java]:
```suggestion
            try (DBCSession session = openSession(monitor, DBCExecutionPurpose.UTIL, "Run extra init queries")) {
                return executeExtraInitQueries(session);
            }
```

File: plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeExecutionContext.java [hunk 6]
```diff
@@ -204,4 +203,26 @@ private void setActiveSchema(DBRProgressMonitor monitor, @NotNull String schemaN
...
+                query,
+                false,
+                false,
+                false))
+            {
+                dbStat.executeStatement();
+            }
+        } catch (Exception e) {
+            log.warn("Error executing initial query.");
+        }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeExecutionContext.java]:
```suggestion
        try (DBCStatement dbStat = session.prepareStatement(DBCStatementType.SCRIPT, query, false, false, false)) {
            dbStat.executeStatement();
        } catch (Exception e) {
            log.warn("Error executing initial query", e);
        }
```

**Review by ShadelessFox — COMMENTED** [review]
Do we really want to execute extra queries if only there were no bootstrap init queries?

File: plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeExecutionContext.java [hunk 7]
```diff
@@ -204,4 +203,21 @@ private void setActiveSchema(DBRProgressMonitor monitor, @NotNull String schemaN
...
+        // JDBC Driver throws java.lang.NoClassDefFoundError for class RootAllocator in Arrow Library, while trying to run a select query.
+        // So this is official workaround for systems with Java 17 (and maybe earlier) and if Arrow library installation is not possible
+        // or if the Operating System is not supported by Arrow.
+        String query = "ALTER SESSION SET JDBC_QUERY_RESULT_FORMAT='JSON'";
+        try {
+            try (DBCStatement dbStat = session.prepareStatement(DBCStatementType.SCRIPT, query, false, false, false))
+            {
+                dbStat.executeStatement();
+            }
+        } catch (Exception e) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.snowflake.core/src/org/jkiss/dbeaver/ext/snowflake/model/SnowflakeExecutionContext.java]:
```suggestion
        try (DBCStatement dbStat = session.prepareStatement(DBCStatementType.SCRIPT, query, false, false, false)) {
            dbStat.executeStatement();
        } catch (Exception e) {
```

**Review by ShadelessFox — APPROVED** [review]

**Review by E1izabeth — APPROVED** [review]

**Comment by LonwoLonwo** [comment]
It was a good, but incorrect solution.

**Pull request closed without merging**